tls/s2n_cipher_suites.c

```diff
@@ -1155,17 +1178,186 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
     return 0;
 }
 
-static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
+/* This function returns the client index for a cipher *match. If a matching cipher could not be found, then -1 is returned. */
+static int s2n_wire_index_of_cipher_suite(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
 {
+    int cipher_not_found = -1;
     for (uint32_t i = 0; i < count; i++) {
         const uint8_t *theirs = wire + (i * cipher_suite_len) + (cipher_suite_len - S2N_TLS_CIPHER_SUITE_LEN);
 
-        if (!memcmp(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
-            return 1;
+        if (s2n_constant_time_equals(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
+            return i;
         }
     }
 
-    return 0;
+    return cipher_not_found;
+}
+
+static bool s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
+{
+    int server_index = s2n_wire_index_of_cipher_suite(match, wire, count, cipher_suite_len);
+    if (server_index < 0) {
+        return false;
+    }
+    return true;
+}
+
+/* Checks that a mutually-supported cipher suite can be used in this connection */
+static bool s2n_cipher_suite_match_is_valid(struct s2n_connection* conn, struct s2n_cipher_suite* potential_match) {
+    /* Never use TLS1.3 ciphers on a pre-TLS1.3 connection, and vice versa */
+    if ((conn->actual_protocol_version >= S2N_TLS13) != (potential_match->minimum_required_tls_version >= S2N_TLS13)) {
+        return false;
+    }
+
+    /* Skip the suite if we don't have an available implementation */
+    if (!potential_match->available) {
+        return false;
+    }
+
+    /* Make sure the cipher is valid for available certs */
+    if (s2n_is_cipher_suite_valid_for_auth(conn, potential_match) != S2N_SUCCESS) {
+        return false;
+    }
+
+    /* TLS 1.3 does not include key exchange in cipher suites */
+    if (potential_match->minimum_required_tls_version < S2N_TLS13) {
+        /* If the kex is not supported continue to the next candidate */
+        bool kex_supported = false;
+        if (s2n_result_is_error(s2n_kex_supported(potential_match, conn, &kex_supported))) {
+            return false;
+        }
+
+        if (!kex_supported) {
+            return false;
+        }
+
+        /* If the kex is not configured correctly continue to the next candidate */
+        if (s2n_result_is_error(s2n_configure_kex(potential_match, conn))) {
+            return false;
+        }
+    }
```

Comment:
If I just added the pointer null checks as you suggested above, then this function doesn't return an error code. Wouldn't it be simpler to just return a bool instead of setting pointers in the argument

Reply:
I'm suggesting it because you're calling methods that do return results / error codes. If you make this method also return a result, this whole section becomes:

```suggestion
    RESULT_GUARD_POSIX(s2n_is_cipher_suite_valid_for_auth(conn, potential_match));

    /* TLS 1.3 does not include key exchange in cipher suites */
    if (potential_match->minimum_required_tls_version < S2N_TLS13) {
        bool kex_supported = false;
        RESULT_GUARD(s2n_kex_supported(potential_match, conn, &kex_supported));
        RESULT_ENSURE(kex_supported, S2N_ERR_CIPHER_NOT_SUPPORTED);
        RESULT_GUARD(s2n_configure_kex(potential_match, conn));
    }
```